File: src/discord_translate_overlay/translation/mock.py
from __future__ import annotations

from ..models import Language
from .base import Translator


class OriginalTranslator(Translator):
    display_name = "원문 표시"
    cache_namespace = "original:v1"

    def translate(self, text: str, source: Language, target: Language) -> str:
        return text


class MockTranslator(Translator):
    display_name = "Mock (테스트)"
    cache_namespace = "mock:v1"

    def translate(self, text: str, source: Language, target: Language) -> str:
        if source == target:
            return text
        return f"[{target.value}] {text}"

File: src-tauri/src/translation/mock.rs
use crate::language::Language;

use super::Translator;

pub struct OriginalTranslator;

impl Translator for OriginalTranslator {
    fn display_name(&self) -> &str {
        "원문 표시"
    }

    fn cache_namespace(&self) -> &str {
        "original:v1"
    }

    fn translate(
        &mut self,
        text: &str,
        _source: Language,
        _target: Language,
    ) -> Result<String, String> {
        Ok(text.to_string())
    }
}

pub struct MockTranslator;

impl Translator for MockTranslator {
    fn display_name(&self) -> &str {
        "Mock (테스트)"
    }

    fn cache_namespace(&self) -> &str {
        "mock:v1"
    }

    fn translate(
        &mut self,
        text: &str,
        source: Language,
        target: Language,
    ) -> Result<String, String> {
        if source == target {
            Ok(text.to_string())
        } else {
            Ok(format!("[{}] {text}", target.code()))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{MockTranslator, Translator};
    use crate::language::Language;

    #[test]
    fn mock_translator_preserves_same_language_and_marks_targets() {
        let mut translator = MockTranslator;
        assert_eq!(
            translator
                .translate("hello", Language::English, Language::English)
                .unwrap(),
            "hello"
        );
        assert_eq!(
            translator
                .translate("hello", Language::English, Language::Korean)
                .unwrap(),
            "[ko] hello"
        );
    }
}
